distributed/http/scheduler/prometheus/core.py

```diff
@@ -52,6 +52,13 @@ def collect(self) -> Iterator[GaugeMetricFamily | CounterMetricFamily]:
         )
         yield worker_states
 
+        if self.server.monitor.monitor_gil_contention:
+            yield GaugeMetricFamily(
```

Comment:
Could we make this into a cumulative metric instead, and use a `Counter` instead of a `Gauge`? With a Gauge, we'll just know what the contention metric happened to be at a particular moment in time every 5s, but not how it's moved around over the prior 5s window.

If we increment `SystemMonitor._cumulative_gil_contention` every time, and expose it as a `Counter`, then the `rate` of that with Prometheus should show us the same thing, but with better sensitivity to what happens in between scrapes.

Reply:
I don't remember fully, but I think @fjetter suggested we do a Guage first then add a Counter on in another PR. I could be mistaken.

Reply:
Sure, I'd be interested to hear the reasoning for separating that, since it's probably a 2-3 line change. I'm not sure of what the benefits could be of a gauge over a counter, but maybe I'm not aware of something? @ntabris do you have any ideas?

Reply:
I was just going to throw in the `Counter` metric as you suggested. However, [according to the Prometheus docs](https://prometheus.io/docs/concepts/metric_types/#counter):
> Do not use a counter to expose a value that can decrease. For example, do not use a counter for the number of currently running processes; instead use a gauge.

Given that the GIL contention can go up/down during this 5s scraping intervals, I'm not sure how to 'convert' it to a counter given that statement. Is [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) more fitting?

Reply:
I think `Counter` makes sense as it takes into account the prometheus polling rate.

Instead of setting a `Gauge` to be equal to `self.server.monitor._last_gil_contention` you would add a new attribute like `self.server.monitor._cumulative_gil_contention` which would be added to every time `_last_gil_contention` is updated, then point a `Counter` to that.

Reply:
So is `_cumulative_gil_contention` type a list to add to? Otherwise I don't see how adding floats of gil contention to a cumulative total between scrapes gives a useful value. Example measure cumulative values `[0.2, 0.2, 0.2, 0.2, 0.2]` summing to  `1.0`  as a counter is different meaning from `[1.0, 0., 0., 0., 0.]` summing to 1.0. (sorry if I'm being dense... :sweat_smile: )

Reply:
And to the docs, explicitly saying not to use a value that can decrease, which the GIL metric will certainly do confuses me.

Reply:
With a counter, you know both the change in value and the time since last value, so you can calculate the average for the polling interval. There's a lot averages don't tell you, but they're still often more informative than samples where the sampling rate is low relative to how often data being sampled changes.

Reply:
> Example measure cumulative values `[0.2, 0.2, 0.2, 0.2, 0.2]` summing to `1.0` as a counter is different meaning from `[1.0, 0., 0., 0., 0.]` summing to 1.0

Yeah, they certainly are different—summary statistics are always misleading! However, 0.2 is probably a better description of the second example that 0.0 (what you'd get from a gauge). Likewise, for `[1.0, 1.0, 1.0, 1.0, 0.2]` for example, 0.84 is probably more accurate than 0.2. We'll never get full temporal resolution with Prometheus, but the average over the last 5s is probably slightly more useful than the instantaneous value.

Especially with something like GIL contention, which I would imagine to be relatively "spiky"—it's probably pretty high-frequency given how low the default switch interval is compared to our 5s sampling rate—I'm hoping this is more useful. You could imagine multiple big spikes within a 5s range that would be lost otherwise.

Reply:
Thanks for the clarification, I appreciate it. Would it be worth it to add/use the histogram metric mentioned earlier?